File: Programmers/DP1.cpp
#include <string>
#include <set>

using namespace std;

int divide(int a, int b) {
    if (!b) return -1;
    else return a / b;
}

int solution(int N, int number) {
    int answer = 9;
    set<int> arr[8];

    for (int i = 0; i < 8; i++) arr[i] = set<int>();

    for (int i = 0; i < 8; i++)
    {
        string tmp = "";

        for (int j = 0; j < i + 1; j++) tmp = tmp + to_string(N);    
        
        arr[i].insert(atoi(tmp.c_str()));

        for (int j = 0; j < i / 2; j++)
        {
            for (auto k = arr[j].begin(); k != arr[j].end(); k++)
            {
                for (auto l = arr[i - j].begin(); l != arr[i - j].end(); l++)
                {
                    int tmp = divide(*k, *l);

                    arr[i].insert(*k + *l);
                    arr[i].insert(*k - *l);
                    arr[i].insert(*k * *l);

                    if (tmp != -1) arr[i].insert(tmp);
                }
            }
        }
    }

    for (int i = 0; i < 8; i++)
    {
        for (auto j = arr[i].begin(); j != arr[i].end(); j++)
        {
            if (*j == number)
            {
                return i + 1;
            }   
        }
    }
    
    return -1;
}

File: Programmers/DP1.py
def solution(N, number):
    answer = -1
    dp = [set() for _ in range(9)]
    
    concat_N = N
    
    for num_cnt in range(1, 9):
        dp[num_cnt].add(concat_N)
        concat_N = concat_N * 10 + N
        for first in range(1, num_cnt):
            second = num_cnt - first
            for val1 in dp[first]:
                for val2 in dp[second]:
                    dp[num_cnt].add(val1 + val2)
                    dp[num_cnt].add(val1 - val2)
                    dp[num_cnt].add(val1 * val2)
                    if val2 != 0:
                        dp[num_cnt].add(val1 // val2)
        
        if number in dp[num_cnt]:
            answer = num_cnt
            break
    
    return answer